executor/common_kvm_arm64_syzos.h

```diff
@@ -79,3 +92,37 @@ GUEST_CODE static void guest_uexit(uint64 exit_code)
 	volatile uint64* ptr = (volatile uint64*)ARM64_ADDR_UEXIT;
 	*ptr = exit_code;
 }
+
+#define GET_OP_FROM_REG(reg, field) \
+	((reg & KVM_REG_ARM64_SYSREG_##field##_MASK) >> KVM_REG_ARM64_SYSREG_##field##_SHIFT)
+
+// Generate an `MSR register, x0` instruction based on the register ID.
+GUEST_CODE static uint32 reg_to_msr(uint64 reg)
+{
+	uint32 opcode = 0xd5100000;
+	// Op0 is 2 bits, but the upper bit is always set to 1 by the opcode.
+	uint32 Op0 = GET_OP_FROM_REG(reg, OP0);
+	uint32 Op1 = GET_OP_FROM_REG(reg, OP1);
+	uint32 CRn = GET_OP_FROM_REG(reg, CRN);
+	uint32 CRm = GET_OP_FROM_REG(reg, CRM);
+	uint32 Op2 = GET_OP_FROM_REG(reg, OP2);
+	return opcode | (Op0 << 19) | (Op1 << 16) | (CRn << 12) | (CRm << 8) | (Op2 << 5);
```

Comment:
Aren't these 19/16/12/8/5 consts hardcode the same KVM_REG_ARM64_SYSREG_##field##_SHIFT values we try to avoid hardcoding above? Both complicating code and still hardcoding values does not make sense.

Are these correct shifts? KVM_REG_ARM64_SYSREG_OP0_SHIFT is 14. The register index seems to take low 2 bytes.

It looks like this can be reduced to:
`return 0xd5100000 | (reg & 0xffff)`
At which point it probably can be just inlined in guest_handle_msr.

Reply:
> Aren't these 19/16/12/8/5 consts hardcode the same KVM_REG_ARM64_SYSREG_##field##_SHIFT values we try to avoid hardcoding above? Both complicating code and still hardcoding values does not make sense.

No, these are the values from the MSR encoding: https://www.scs.stanford.edu/~zyedidia/arm64/msr_reg.html

But you are right that for MSR these values follow each other, so the instruction can be encoded as `0xd5100000 | ((reg & 0xffff) << 5)`